projects/mysadna/src/app/services/firebase.service.ts

```diff
@@ -13,37 +13,106 @@ interface FirebaseElement {
 @Injectable()
 export class FirebaseService {
   private protobin: AngularFirestoreCollection<FirebaseElement>;
+  private userList: AngularFirestoreCollection<FirebaseElement>;
+  private projectList: AngularFirestoreCollection<FirebaseElement>;
+  data: Data;
+  users: User[] = [];
+  projects: Project[] = [];
+  tempUser: User;
+  tempProject: Project;
 
   constructor(
     private db: AngularFirestore,
     private encodingService: EncodingService,
   ) {
     this.protobin = this.db.collection('protobin');
+    this.projectList = this.db.collection('protobin/data/project-list');
+    this.userList = this.db.collection('protobin/data/user-list');
   }
 
-  getData(): Observable<Data> {
-    return this.protobin
-      .doc('data')
-      .snapshotChanges()
-      .pipe(
-        map(action => {
-          const firebaseElement = action.payload.data() as FirebaseElement;
-          if (firebaseElement === undefined) {
-            // Element not found
-            return;
-          }
-          return this.convertFirebaseElementToData(firebaseElement);
-        })
-      );
+  InitData(): void {
+    this.data = new Data();
+    this.getUserList().subscribe(users => {
+      this.data.setUserList(users);
+    });
+    this.getProjectList().subscribe(projects =>  {
+      this.data.setProjectList(projects);
+    });
   }
 
-  private convertFirebaseElementToData(firebaseElement: FirebaseElement): Data {
+  getUserList(): Observable<User[]> {
+    return this.userList.snapshotChanges().pipe(
+      map(action => action.map(a => {
+        const firebaseElement = a.payload.doc.data() as FirebaseElement;
+
+        if (firebaseElement === undefined) {
+          // Element not found
+          return;
+        }
+
+        this.tempUser = this.convertFirebaseElementToUser(firebaseElement);
+
+        if (this.users.find(user => user.getUserId() === this.tempUser.getUserId())) {
+          const userIndex = this.users.findIndex(user => user.getUserId() === this.tempUser.getUserId());
+          this.users[userIndex] = this.tempUser;
+        } else {
+          this.users.push(this.tempUser);
+        }
+        return this.tempUser;
+      }))
+    );
+  }
+
+  getProjectList(): Observable<Project[]> {
+    return this.projectList.snapshotChanges().pipe(
+      map(action => action.map(a => {
+        const firebaseElement = a.payload.doc.data() as FirebaseElement;
+
+        if (firebaseElement === undefined) {
+          // Element not found
+          return;
+        }
+
+        this.tempProject = this.convertFirebaseElementToProject(firebaseElement);
+
+        if (this.projects.find(project => project.getProjectId() === this.tempProject.getProjectId())) {
+          const projectIndex = this.projects.findIndex(project => project.getProjectId() === this.tempProject.getProjectId());
+          this.projects[projectIndex] = this.tempProject;
```

Comment:
I thought about it, but `User`s and `Project`s are completely different classes.

Reply:
I'll show you

Reply:
How about that:
```typescript
  getUserList(): Observable<User[]> {
    return this.userList.snapshotChanges().pipe(
      map(action => action.map(a => {
        const firebaseElement: FirebaseElement = a.payload.doc.data();
        return User.deserializeBinary(this.getBinary(firebaseElement));
      }))
    );
  }

  getProjectList(): Observable<Project[]> {
    return this.projectList.snapshotChanges().pipe(
      map(action => action.map(a => {
        const firebaseElement: FirebaseElement = a.payload.doc.data();
        return Project.deserializeBinary(this.getBinary(firebaseElement));
      }))
    );
  }

  // Converts firebaseElement to binary
  private getBinary(firebaseElement: FirebaseElement): Uint8Array {
    const binary: Uint8Array = this.encodingService
      .decodeBase64StringToUint8Array(firebaseElement.proto);
    return binary;
  }
```

Reply:
I'll show you

Reply:
I've tried to code a universal method for both Users and Projects, but it didn't work since I couldn't know the returned object's type.

Reply:
Yes, you're right. We need the two methods, because they have different return type.
But we can simplify content of the methods and also `convertFirebaseElementTo` methods.
Actually, if you saw my comment in you email, you already know what I mean.

Reply:
I didn't really understand from the code in the email.